File: apps/flutter/lib/features/debts/domain/entities/debt_transaction.dart
class DebtTransaction {
  final String id;
  final String debtId;
  final DateTime date;
  final String type;
  final double amount;
  final String? accountId;
  final String? note;

  const DebtTransaction({
    required this.id,
    required this.debtId,
    required this.date,
    required this.type,
    required this.amount,
    required this.accountId,
    required this.note,
  });

  factory DebtTransaction.fromJson(Map<String, dynamic> json) {
    return DebtTransaction(
      id: (json["debtTransactionId"] ?? json["id"]) as String,
      debtId: json["debtId"] as String,
      date: DateTime.parse(json["date"] as String),
      type: json["type"] as String,
      amount: (json["amount"] as num?)?.toDouble() ?? 0,
      accountId: json["accountId"] as String?,
      note: json["note"] as String?,
    );
  }
}

File: apps/api/src/models/debt_transaction.ts
import { AggregateRoot } from "@abejarano/ts-mongodb-criteria";
import { createMongoId } from "./shared/mongo_id";

export enum DebtTransactionType {
  Charge = "charge",
  Payment = "payment",
  Fee = "fee",
  Interest = "interest",
}

export type DebtTransactionPrimitives = {
  id?: string;
  debtTransactionId: string;
  userId: string;
  debtId: string;
  date: Date;
  type: DebtTransactionType;
  amount: number;
  accountId?: string;
  categoryId?: string | null;
  note?: string | null;
  createdAt: Date;
  updatedAt?: Date;
};

export type DebtTransactionCreateProps = {
  userId: string;
  debtId: string;
  date?: Date;
  type: DebtTransactionType;
  amount: number;
  accountId?: string;
  categoryId?: string | null;
  note?: string | null;
};

export class DebtTransaction extends AggregateRoot {
  private readonly props: DebtTransactionPrimitives;

  private constructor(props: DebtTransactionPrimitives) {
    super();
    this.props = props;
  }

  static create(props: DebtTransactionCreateProps): DebtTransaction {
    const now = new Date();
    return new DebtTransaction({
      debtTransactionId: createMongoId(),
      userId: props.userId,
      debtId: props.debtId,
      date: props.date ?? now,
      type: props.type,
      amount: props.amount,
      accountId: props.accountId,
      categoryId: props.categoryId ?? null,
      note: props.note,
      createdAt: now,
      updatedAt: now,
    });
  }

  static fromPrimitives(primitives: DebtTransactionPrimitives): DebtTransaction {
    return new DebtTransaction(primitives);
  }

  toPrimitives(): DebtTransactionPrimitives {
    return this.props;
  }

  getId(): string {
    return this.props.id ?? this.props.debtTransactionId;
  }

  getDebtTransactionId(): string {
    return this.props.debtTransactionId;
  }
}
